### PR TITLE
fix(java): size wire strings by UTF-8 byte length, not char count

### PR DESCRIPTION
**Pull request by mmodzelewski**
Names, keys and passwords with non-ASCII characters went out with a
length prefix counted in UTF-16 chars while the bytes came from the
platform default charset, so the prefix and payload disagreed and the
server misread the frame. The 255 cap was also checked in chars, so a
short name could still overflow the u8 prefix, and uncapped strings
wrapped it silently.

Every wire string is now encoded as UTF-8 once and prefixed with its
byte length. The serializer rejects empty or over-long strings and
names the field. Identifier caches the encoded name, and Partitioning
checks the value shape the server expects for each kind. The SDK
version is cut at a code point boundary. Message.of, the example
consumer and the javadoc all use UTF-8 explicitly.

The login payload no longer carries version and context strings; the
VSR codec drops them and sends the SDK version itself. The codec now
validates fields before allocating, closing a pooled-buffer leak on
an empty username. Usernames, passwords and token names are checked
against the server's own bounds before the round trip.

Message keys are hashed as UTF-8 bytes, so on a JVM whose default
charset is not UTF-8 a non-ASCII key may land on a different
partition than before. Needs a 0.9.0 release note. Non-ASCII names
are tested end to end over TCP only; the HTTP client does not
percent-encode paths yet.

Refs #4056

### CODE REVIEW

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java [hunk 1]
```diff
@@ -68,13 +80,21 @@ public String getName() {
         return name;
     }
 
+    /**
+     * The name as UTF-8 wire bytes, encoded once at construction; {@code null} for a numeric
+     * identifier. The array is shared rather than copied, so callers must not modify it.
+     */
+    @Nullable public byte[] getEncodedName() {
```
Comment (spetz) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java:87]: Consider returning a defensive copy or a read-only view. This exposes the cached serialization bytes: `id.getEncodedName()[0] = 'b'` changes `StreamId.of("foo")` into `"boo"` on the wire while `getName()`, logging, and routing-cache keys still identify `"foo"`. The Javadoc documents the restriction, but enforcing it would preserve identifier consistency

File: foreign/java/java-sdk/src/test/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodecTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    @ParameterizedTest
+    @NullSource
+    @EmptySource
+    void sdkVersionFieldFallsBackToUnknownWhenVersionIsMissing(String version) {
+        assertThat(VsrLoginCodec.sdkVersionField(version)).isEqualTo("unknown".getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Test
+    void sdkVersionFieldTruncatesOnCodePointBoundaryWithinU8Prefix() {
```
Comment (spetz) [inline foreign/java/java-sdk/src/test/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodecTest.java:47]: Please add surrogate-pair boundary cases. The existing `é` test covers multibyte UTF-8 but not supplementary code points. `"a".repeat(251) + "\uD83D\uDE00"` should encode to 255 bytes. With 252 leading ASCII characters, the result should contain only those 252 bytes. This verifies that truncation preserves complete code points

**Review by spetz — CHANGES_REQUESTED** [review]

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/UsersTcpClient.java [hunk 3]
```diff
@@ -155,20 +157,23 @@ public CompletableFuture<IdentityInfo> login(String username, String password) {
         return routingHook.loginOnLeader(() -> loginWithoutRedirect(username, password));
     }
 
+    static ByteBuf loginPayload(String username, String password, String version, String context) {
+        byte[] versionBytes = version.getBytes(StandardCharsets.UTF_8);
+        byte[] contextBytes = context.getBytes(StandardCharsets.UTF_8);
+        var payload = Unpooled.buffer();
+        payload.writeBytes(toBytes(username));
+        payload.writeBytes(toBytes(password));
+        payload.writeIntLE(versionBytes.length);
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/UsersTcpClient.java]: warning: these bytes never reach the wire - `VsrLoginCodec.rewriteUserLogin` reads username and password only, then hardcodes `writeIntLE(0)` for the context. drop version and context here, or thread `context` through the codec.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java [hunk 4]
```diff
@@ -23,8 +23,20 @@
...
 import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
 
 public record Partitioning(PartitioningKind kind, byte[] value) {
+
+    /** Server-side cap on a messages key, in encoded bytes, matching its u8 length prefix. */
+    public static final int MAX_MESSAGES_KEY_LENGTH = 255;
+
+    public Partitioning {
+        if (value.length > MAX_MESSAGES_KEY_LENGTH) {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java]: warning: this only checks the ceiling, and it checks it for every kind. an empty `MessagesKey` and a `PartitionId` that isn't 4 bytes both pass here and get rejected by the server - switch on `kind` instead.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java [hunk 5]
```diff
@@ -38,10 +50,15 @@ public static Partitioning partitionId(Long id) {
     }
 
     public static Partitioning messagesKey(String key) {
-        if (key == null || key.isBlank() || key.length() > 255) {
-            throw new IggyInvalidArgumentException("Key must be non-empty and less than 255 characters long");
+        if (key == null || key.isBlank()) {
+            throw new IggyInvalidArgumentException("Key must be non-empty");
+        }
+        byte[] encoded = key.getBytes(StandardCharsets.UTF_8);
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java]: warning: this changes message-key routing - `MessagesTcpClient` hashes these bytes client-side, so a non-ASCII key lands on a different partition after the upgrade on any JVM whose default charset isn't UTF-8. correct change, but it needs a release note.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/serde/BytesSerializer.java [hunk 6]
```diff
@@ -63,6 +63,9 @@ public final class BytesSerializer {
      */
     private static final int MAX_HEADER_FIELD_LENGTH = 255;
 
+    /** Bound on a u8-length-prefixed wire string, in encoded bytes. */
+    private static final int MAX_U8_STRING_LENGTH = 255;
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/serde/BytesSerializer.java:65]: warning: 255 is the wire cap, but the server bounds username at 3..50, password at 3..100 and PAT name at 3..30. a 200-byte username passes here and then dies at the server - bound them at the call sites.

File: foreign/java/java-sdk/src/test/java/org/apache/iggy/message/MessageTest.java [hunk 7]
```diff
@@ -72,6 +73,14 @@ void ofCreatesMessageWhenGivenMessageHeaderPayloadAndNullUserHeaders() {
         assertThat(message.userHeaders().size()).isEqualTo(0);
     }
 
+    @Test
+    void ofEncodesStringPayloadAsUtf8() {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/test/java/org/apache/iggy/message/MessageTest.java:77]: warning: this passes against pre-PR code on any UTF-8-default JVM, and the Test task sets no `defaultCharacterEncoding`, so CI never runs the case it's named for. same at `PartitioningTest.java:81`. fork the test JVM with `file.encoding=ISO-8859-1`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/serde/BytesSerializer.java [hunk 8]
```diff
@@ -210,9 +214,12 @@ public static ByteBuf toBytes(TopicPermissions permissions) {
     }
 
     public static ByteBuf toBytes(String value) {
-        int bufferLength = 1 + value.length();
-        ByteBuf buffer = Unpooled.buffer(bufferLength);
         byte[] stringBytes = value.getBytes(StandardCharsets.UTF_8);
+        if (stringBytes.length > MAX_U8_STRING_LENGTH) {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/serde/BytesSerializer.java]: nit: seven fields funnel through here - stream, topic and group names, PAT name, username, password, PAT token - so this message tells the caller nothing about which one broke. `checkFieldLength` below takes a field label for exactly that reason.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/serde/BytesSerializer.java [hunk 9]
```diff
@@ -210,9 +214,12 @@ public static ByteBuf toBytes(TopicPermissions permissions) {
...
+            throw new IggyInvalidArgumentException("String must be at most " + MAX_U8_STRING_LENGTH
+                    + " bytes when UTF-8 encoded, got " + stringBytes.length);
+        }
+        ByteBuf buffer = Unpooled.buffer(1 + stringBytes.length);
         buffer.writeByte(stringBytes.length);
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/serde/BytesSerializer.java:215]: nit: no lower bound - `toBytes("")` still writes `[0x00]`, and the server rejects a zero-length wire name, so `createStream("")` builds a frame it always refuses. pre-existing, and adding a floor means flipping `shouldSerializeEmptyString`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java (hunk 1 above)
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java]: nit: new public API on a public abstract class that hands back the live array. a caller that mutates it keeps the length prefix valid and silently sends a different name than `getName()` reports.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java [hunk 10]
```diff
@@ -23,8 +23,20 @@
...
 
 import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
 
 public record Partitioning(PartitioningKind kind, byte[] value) {
+
+    /** Server-side cap on a messages key, in encoded bytes, matching its u8 length prefix. */
+    public static final int MAX_MESSAGES_KEY_LENGTH = 255;
+
+    public Partitioning {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java:35]: nit: `new Partitioning(kind, null)` throws a bare NPE at the length check instead of `IggyInvalidArgumentException`. same shape at `partitionId(null)`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodec.java [hunk 11]
```diff
@@ -94,15 +103,22 @@ static long readSessionEpoch(ByteBuf registerBody) {
...
+     * always fits its u8 prefix and still decodes as UTF-8 on the server.
+     */
+    static byte[] sdkVersionField(String version) {
+        String value = version == null || version.isEmpty() ? UNKNOWN_SDK_VERSION : version;
+        ByteBuffer encoded = ByteBuffer.allocate(MAX_SHORT_FIELD_LENGTH);
+        StandardCharsets.UTF_8
+                .newEncoder()
+                .onMalformedInput(CodingErrorAction.REPLACE)
+                .onUnmappableCharacter(CodingErrorAction.REPLACE)
+                .encode(CharBuffer.wrap(value), encoded, true);
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodec.java]: nit: the encoder contract wants `flush(out)` after `encode(in, out, true)`. it works today only because the UTF-8 encoder's `implFlush` is a no-op - append `.flush(encoded)`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodec.java [hunk 12]
```diff
@@ -94,15 +103,22 @@ static long readSessionEpoch(ByteBuf registerBody) {
     private static void writeVersionInfo(ByteBuf body) {
         body.writeIntLE(PROTOCOL_VERSION);
         writeShortField(body, SDK_NAME.getBytes(StandardCharsets.UTF_8));
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodec.java]: nit: `SDK_NAME` and the version are both constants, re-encoded on every login and reconnect. hoist them to `private static final byte[]`, keeping `sdkVersionField` package-visible for the test.

File: foreign/java/java-sdk/src/test/java/org/apache/iggy/client/blocking/tcp/BytesSerializerTest.java [hunk 13]
```diff
@@ -239,6 +267,44 @@ void shouldSerializeStringIdentifier() {
...
+            assertThat(result.readableBytes()).isEqualTo(0);
+        }
+
+        @ParameterizedTest
+        @CsvSource({
+            "café,       02 05 63 61 66 C3 A9",
+            "naïve-café, 02 0C 6E 61 C3 AF 76 65 2D 63 61 66 C3 A9",
+            "日本語,     02 09 E6 97 A5 E6 9C AC E8 AA 9E",
+        })
+        void shouldMatchServerWireFormatForNonAsciiNames(String name, String expectedHex) {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/test/java/org/apache/iggy/client/blocking/tcp/BytesSerializerTest.java]: nit: the name says server wire format, but the body compares hand-written hex with nothing tying it to the Rust encoder. rename to `shouldMatchExpectedWireLayoutForNonAsciiNames`, or move the parity claim into `bdd/`.

File: foreign/java/java-sdk/src/test/java/org/apache/iggy/client/blocking/tcp/BytesSerializerTest.java [hunk 14]
```diff
@@ -239,6 +267,44 @@ void shouldSerializeStringIdentifier() {
             result.readBytes(nameBytes);
             assertThat(new String(nameBytes)).isEqualTo("test-stream");
```
Comment (hubcio) [inline foreign/java/java-sdk/src/test/java/org/apache/iggy/client/blocking/tcp/BytesSerializerTest.java]: nit: `new String(nameBytes)` decodes on the platform default, the same conflation this PR is removing. pass `StandardCharsets.UTF_8`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java [hunk 15]
```diff
@@ -38,10 +50,15 @@ public static Partitioning partitionId(Long id) {
     }
 
     public static Partitioning messagesKey(String key) {
-        if (key == null || key.isBlank() || key.length() > 255) {
-            throw new IggyInvalidArgumentException("Key must be non-empty and less than 255 characters long");
+        if (key == null || key.isBlank()) {
+            throw new IggyInvalidArgumentException("Key must be non-empty");
+        }
+        byte[] encoded = key.getBytes(StandardCharsets.UTF_8);
+        if (encoded.length > MAX_MESSAGES_KEY_LENGTH) {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java]: simplification: the compact constructor applies this same bound four lines down. drop the check here and let the constructor throw - `hasMessageContaining("300")` still passes.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java [hunk 16]
```diff
@@ -23,10 +23,15 @@
 import org.apache.iggy.exception.IggyInvalidArgumentException;
 
 import javax.annotation.Nullable;
+import java.nio.charset.StandardCharsets;
 
 public abstract class Identifier {
 
+    /** Server-side cap on a wire name, in UTF-8 bytes, matching its u8 length prefix. */
+    public static final int MAX_NAME_LENGTH = 255;
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java]: simplification: nothing outside this file uses `MAX_NAME_LENGTH`, and the same holds for `Partitioning.MAX_MESSAGES_KEY_LENGTH`. both are new, so making them private breaks nothing.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/ConsumerGroupsTcpClient.java [hunk 17]
```diff
@@ -106,11 +106,13 @@ public CompletableFuture<ConsumerGroupDetails> createConsumerGroup(
             StreamId streamId, TopicId topicId, String name) {
         var streamIdBytes = BytesSerializer.toBytes(streamId);
         var topicIdBytes = BytesSerializer.toBytes(topicId);
-        var payload = Unpooled.buffer(1 + streamIdBytes.readableBytes() + topicIdBytes.readableBytes() + name.length());
+        var nameBytes = BytesSerializer.toBytes(name);
+        var payload = Unpooled.buffer(
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/ConsumerGroupsTcpClient.java]: simplification: `toBytes(streamId)` followed by two `writeBytes` calls does the same thing without the manual sizing, matching `:61` and `:84` in this file. same shape at `StreamsTcpClient.java:104`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/StreamsTcpClient.java [hunk 18]
```diff
@@ -102,12 +99,12 @@ public CompletableFuture<List<StreamBase>> getStreams() {
...
-        var payload = Unpooled.buffer(payloadSize + idBytes.capacity());
+        var nameBytes = BytesSerializer.toBytes(name);
+        var payload = Unpooled.buffer(idBytes.readableBytes() + nameBytes.readableBytes());
 
         payload.writeBytes(idBytes);
-        payload.writeBytes(BytesSerializer.toBytes(name));
+        payload.writeBytes(nameBytes);
         // Trailing options block. Streams have no catalog keys yet, so the
         // server rejects every key; the empty block is the extension point.
         payload.writeBytes(BytesSerializer.toBytes(Map.<HeaderKey, HeaderValue>of()));
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/StreamsTcpClient.java]: simplification: `toBytes` of an empty map returns `EMPTY_BUFFER`, so this writes nothing. the comment above already marks the extension point - same at `UsersTcpClient.java:125`.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodec.java [hunk 19]
```diff
@@ -94,15 +103,22 @@ static long readSessionEpoch(ByteBuf registerBody) {
...
-        String version = IggyVersion.getInstance().getVersion();
-        if (version == null || version.isEmpty()) {
-            return "unknown";
-        }
-        return version.length() > 255 ? version.substring(0, 255) : version;
+    /**
+     * An over-long version is cut on the encoded bytes at a code point boundary, so the field
+     * always fits its u8 prefix and still decodes as UTF-8 on the server.
+     */
+    static byte[] sdkVersionField(String version) {
```
Comment (hubcio) [inline foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodec.java:115]: simplification: `getBytes(UTF_8)` already replaces unpaired surrogates, so this reduces to encode once, return if it's 255 bytes or under, else walk back off the continuation bytes. drops the flush issue too.

**Review by hubcio — CHANGES_REQUESTED** [review]
Four things that aren't on a changed line:

- **The login third fixes a path with no bug.** `VsrLoginCodec.rewriteUserLogin` reads username and password only and hardcodes `writeIntLE(0)` for the context at `:76`, so those bytes never left the process pre-PR either. `UsersTcpClientPayloadTest` now pins a layout no server sees.
- **Pooled buffer leak at `VsrLoginCodec.java:71`.** `writeShortField` throws on an empty username at `:138`, and `VsrRequestEncoder.java:58` calls the codec outside its try/finally. Reachable via `login("", pw)` - move the length check above the allocation.
- **Breaks a round trip in the examples.** `Message.of` now encodes UTF-8, but its partner `AsyncConsumer.java:315` still decodes on the platform default. Same line in the `MessagesClient.java:62` javadoc.
- **No non-ASCII test reaches a real server.** `MessagesTcpClientTest.java:34` already covers the `metadataLength` sum end to end, but with ASCII names only.

For the description: `toBytes(String)` now rejects over-long usernames, passwords and PAT fields too, and the message-key routing change needs a line in the 0.9.0 notes.

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java (hunk 1 above)
Comment (mmodzelewski) [reply foreign/java/java-sdk/src/main/java/org/apache/iggy/identifier/Identifier.java]: raw bytes array is no longer shared

File: foreign/java/java-sdk/src/test/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodecTest.java (hunk 2 above)
Comment (mmodzelewski) [reply foreign/java/java-sdk/src/test/java/org/apache/iggy/client/async/tcp/vsr/VsrLoginCodecTest.java:54]: added new test cases

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java (hunk 5 above)
Comment (mmodzelewski) [reply foreign/java/java-sdk/src/main/java/org/apache/iggy/message/Partitioning.java]: added info to the PR description

File: foreign/java/java-sdk/src/test/java/org/apache/iggy/message/MessageTest.java (hunk 7 above)
Comment (mmodzelewski) [reply foreign/java/java-sdk/src/test/java/org/apache/iggy/message/MessageTest.java:77]: I think it's not worth it spawning a separate test suite just for it. From JDK 18, UTF-8 is the default. Before 18, ISO-8859-1 would only be set on Windows.

**Review by hubcio — APPROVED** [review]

**Review by spetz — APPROVED** [review]
